services/provider/server/server.go

```diff
@@ -694,21 +703,84 @@ func (s *OCSProviderServer) getOCSSubscriptionChannel(ctx context.Context) (stri
 		klog.Info("unable to find ocs-operator subscription")
 		return "", nil
 	}
-	if subscription.Status.InstalledCSV == "" {
+
+	return s.getLastKnownValidChannel(ctx, subscription)
+}
+
+func (s *OCSProviderServer) getLastKnownValidChannel(ctx context.Context, subscription *opv1a1.Subscription) (string, error) {
+	// use mutex lock for using the package level variables to avoid concurrency issues among rpc calls
+	lastKnownMutex.Lock()
+	defer lastKnownMutex.Unlock()
+
+	installedCSVName := subscription.Status.InstalledCSV
+	// If the installed CSV has not changed no need to revalidate & update the last known values
+	if installedCSVName == lastKnownInstalledCSVname {
+		return lastKnownSubChannelName, nil
+	}
+
+	// Check if the installed CSV of the subscription is in Succeeded phase
+	if installedCSVName == "" {
 		klog.Infof("Subscription %q does not have an installed CSV", subscription.Name)
-		return "", nil
+		return lastKnownSubChannelName, nil
 	}
 	csv := &opv1a1.ClusterServiceVersion{}
-	err = s.client.Get(ctx, client.ObjectKey{Name: subscription.Status.InstalledCSV, Namespace: s.namespace}, csv)
+	err := s.client.Get(ctx, client.ObjectKey{Name: installedCSVName, Namespace: s.namespace}, csv)
 	if err != nil {
-		klog.Errorf("failed to get installed CSV %q: %v", subscription.Status.InstalledCSV, err)
+		klog.Errorf("failed to get installed CSV %q: %v", installedCSVName, err)
 		return "", err
 	}
 	if csv.Status.Phase != opv1a1.CSVPhaseSucceeded {
 		klog.Infof("installed CSV %q is in phase %q, not Succeeded", csv.Name, csv.Status.Phase)
-		return "", nil
+		return lastKnownSubChannelName, nil
 	}
-	return subscription.Spec.Channel, nil
+
+	// Check if the installed CSV is the expected one for the channel from the package manifest
+	packageManifest := &unstructured.Unstructured{}
+	packageManifest.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "packages.operators.coreos.com",
+		Version: "v1",
+		Kind:    "PackageManifest",
+	})
+	err = s.client.Get(ctx, client.ObjectKey{Name: subscription.Spec.Package, Namespace: s.namespace}, packageManifest)
+	if err != nil {
+		klog.Errorf("failed to get PackageManifest %q: %v", subscription.Spec.Package, err)
+		return "", err
+	}
+	channels, found, err := unstructured.NestedSlice(packageManifest.Object, "status", "channels")
+	if err != nil {
+		klog.Errorf("error getting channels from PackageManifest %q: %v", packageManifest.GetName(), err)
+		return "", err
+	}
+	if !found {
+		klog.Infof("channels not found in PackageManifest %q", packageManifest.GetName())
+		return lastKnownSubChannelName, nil
```

Comment:
Why do we consider this a non-error?

Reply:
Now any problem faced during the validation of the channel is an error